store.go

```diff
@@ -3682,3 +3753,30 @@ func (s *store) Free() {
 		}
 	}
 }
+
+func searchLayer(ls drivers.AdditionalLayerStore, d digest.Digest, ref string) (*Layer, string, error) {
+	var elems []string
+	if ls.WithReference {
+		elems = append(elems, base64.StdEncoding.EncodeToString([]byte(ref)))
+	}
+	elems = append(elems, d.Encoded())
+	target := filepath.Join(append([]string{ls.Path}, elems...)...)
+	if _, err := os.Stat(filepath.Join(target, "diff")); err == nil {
```

Comment:
The filesystem accesses here feel weird to me.

If it is the generic code that defines the layout and file format, and the drivers must accept absolute paths to arbitrary locations, why is the driver in charge of maintaining `driver.AdditionalLayerStores()`  at all, and it isn’t a generic c/storage option?

Alternatively, shouldn’t the driver be fully in charge of the filesystem layout, so that the generic code only calls “find”/“use” operations provided by the driver?

I’m leaning towards the second view — why does the generic code care about a `diff` subdirectory at all, when that’s an overlay-driver-specific concept?

Reply:
> Alternatively, shouldn’t the driver be fully in charge of the filesystem layout, so that the generic code only calls “find”/“use” operations provided by the driver?

Thanks for the suggestion. I think it's possible by adding something like the following API to the driver. I'll add this if this idea LGTY.

- `CreateFromAdditionalLayerStore(id, parent string) error` : creates a layer using the diff contents from the additional layer store.
- `AdditionalLayerInfo(d digest.Digest, ref string) (io.ReadCloser, error)` : provides `info` file contents. The store will decode this to `*c/storage.Layer` struct.

Reply:
I’ll defer to c/storage maintainers. (Another design concern: Should the API, in advance, account for multiple formats of the on-demand-created data?)

Reply:
Moved the directory-structure-aware logic into overlay driver. The driver's API is format-agnostic I believe.